main/lsp/LSPTypechecker.cc

```diff
@@ -36,44 +36,54 @@ void sendTypecheckInfo(const LSPConfiguration &config, const core::GlobalState &
 }
 } // namespace
 
-LSPTypechecker::LSPTypechecker(const std::shared_ptr<const LSPConfiguration> &config)
-    : typecheckerThreadId(this_thread::get_id()), config(config) {}
+LSPTypechecker::LSPTypechecker(std::shared_ptr<const LSPConfiguration> config)
+    : typecheckerThreadId(this_thread::get_id()), config(move(config)) {}
 
 void LSPTypechecker::initialize(LSPFileUpdates updates) {
     globalStateHashes = move(updates.updatedFileHashes);
     indexed = move(updates.updatedFileIndexes);
     // Initialization typecheck is not cancelable.
-    auto run = runSlowPath(move(updates), /* cancelable */ false);
-    ENFORCE(!run.canceled);
-    ENFORCE(run.newGS.has_value());
-    gs = move(run.newGS.value());
-    pushDiagnostics(move(run));
+    auto committed = runSlowPath(move(updates), /* cancelable */ false);
+    ENFORCE(committed);
 }
 
 bool LSPTypechecker::typecheck(LSPFileUpdates updates) {
-    sendTypecheckInfo(*config, *gs, SorbetTypecheckRunStatus::Started, updates.canTakeFastPath, {});
-    auto run = runTypechecking(move(updates));
-    auto committed = !run.canceled;
-    vector<core::FileRef> filesTypechecked = run.filesTypechecked;
-    const bool isFastPath = run.updates.canTakeFastPath;
-    commitTypecheckRun(move(run));
+    vector<core::FileRef> filesTypechecked;
+    bool committed = true;
+    const bool isFastPath = updates.canTakeFastPath;
+    sendTypecheckInfo(*config, *gs, SorbetTypecheckRunStatus::Started, isFastPath, {});
+    if (isFastPath) {
+        auto run = runFastPath(move(updates));
```

Comment:
Is it not possible for a fast path to be canceled?

Reply:
Correct. Fast path cannot be canceled.

Reply:
Is there some guarantee that we only do one fast path at a time?  For example, if I'm editing the body of a long HEREDOC, for example, will I spawn a ton of fast path typechecks while I'm editing the body, since my edits are highly unlikely to be modifying any symbols?  (Or is fast path so fast that it doesn't matter at human typing speed?)

Reply:
Resolved in side-conversation.

Reply:
Answer: Yup. Typechecker thread only processes one task at a time.

The answer is still yes with preemption. Preemption, which pauses a slow path to run some task T, can only happen during the slow path. Preemption tasks cannot run slow paths and cannot be preempted themselves.